### PR TITLE
Make index methods async

### PR DESCRIPTION
**Pull request by johnhaley81**
Anytime the index reads/writes to the disk it should be non-blocking. Additionally this will give us thread safety protection since async methods are piped through the `LockManager`

### CODE REVIEW

**Comment by johnhaley81** [comment]
Also updated the examples to both use the async methods and remove the double reading of the index. `repo.openIndex` forces a read of the index so it doesn't make sense to do another one right after the first one happens. I think that was a holdover from an ancient version of libgit2.

**Comment by joshaber** [comment]
💖

**Comment by johnhaley81** [comment]
This should be gtg now. @tbranyen, @srajko, and @joshaber if you wouldn't mind :eyes: on this that would be most appreciated :)

**Comment by tbranyen** [comment]
We need to start using async/await internally. ^_^

**Comment by johnhaley81** [comment]
#874 :smile:

File: generate/input/descriptor.json [hunk 1]
```diff
@@ -992,6 +1021,13 @@
               "isReturn": true
             }
           },
+          "isAsync": true,
+          "return": {
+            "isErrorCode": true
+          }
+        },
+        "git_index_conflict_remove": {
```
Comment (srajko) [inline generate/input/descriptor.json]: If you want to be pedantic about keeping things sorted alphabetically, this one belongs a few entries down

**Pull request merged**